src/coreclr/hosts/corerun/corerun.cpp

```diff
@@ -131,20 +132,36 @@ static string_t build_tpa(const string_t& core_root, const string_t& core_librar
             string_t tmp = pal::build_file_list(dir, ext, [&](const char_t* file)
                 {
                     string_t file_local{ file };
+                    string_t file_name { file }; 
 
                     // Strip the extension.
                     if (pal::string_ends_with(file_local, ext_len, ext))
-                        file_local = file_local.substr(0, file_local.length() - ext_len);
+                        file_name = file_local.substr(0, file_local.length() - ext_len);
 
                     // Return true if the file is new.
-                    return name_set.insert(file_local).second;
+                    return name_path_map.insert({file_name, file_local}).second;
                 });
 
             // Add to the TPA.
             tpa_list << tmp;
         }
     }
 
+    tpa_items->assembly_count = static_cast<uint32_t>(name_path_map.size());
```

Comment:
There is a number of apps out there that host coreclr.dll directly using `TRUSTED_PLATFORM_ASSEMBLIES` and other key/value pairs. We had the direct coreclr.dll hosting documented at learning.microsoft.com for a while. There are scenarios where hosting via host policy is too inflexible.

Are we going to break the existing coreclr.dll hosting model based on `TRUSTED_PLATFORM_ASSEMBLIES` and other key/value pairs or are we keeping it around?

If we are keeping it around, we may want to keep corerun intact so that we get some test coverage for it. corerun is our non-shipping test-only host and it is fine for it to leave some perf on table.

If we are going to break it, it will be a significant breaking change that will require a lot of work to push through.

Reply:
> We do not plan to break it.

Ok, we should think about having some test coverage. `corerun` test host would be a good place for it - keeping it intact should be the easiest option.